### ISSUE TITLE
Monkey patch hangs when using IPython

### ISSUE THREAD

**Issue by isofer** [eventlet/eventlet#395]
Hi guys,
Thanks for this great package.

I found an issue when using IPython: monkey patching just hangs.
It works fine when I use regular python. It only hangs with IPython.

The traceback after I interrupt:
```
Traceback (most recent call last):
  File "/Users/imris/anaconda/envs/test/bin/ipython", line 6, in <module>
    sys.exit(IPython.start_ipython())
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/__init__.py", line 119, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/traitlets/config/application.py", line 658, in launch_instance
    app.start()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/terminal/ipapp.py", line 348, in start
    self.shell.mainloop()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/terminal/interactiveshell.py", line 486, in mainloop
    self.interact()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/terminal/interactiveshell.py", line 469, in interact
    code = self.prompt_for_code()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/terminal/interactiveshell.py", line 363, in prompt_for_code
    pre_run=self.pre_prompt, reset_current_buffer=True)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/prompt_toolkit/interface.py", line 394, in run
    self.eventloop.run(self.input, self.create_eventloop_callbacks())
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/prompt_toolkit/eventloop/posix.py", line 115, in run
    fds = self._ready_for_reading(remaining_timeout)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/prompt_toolkit/eventloop/posix.py", line 193, in _ready_for_reading
    fds = self.selector.select(timeout)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/prompt_toolkit/eventloop/select.py", line 100, in select
    return self._select_selector.select(timeout)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/prompt_toolkit/eventloop/select.py", line 182, in select
    return select.select(self._fds, [], [], timeout)[0]
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/green/select.py", line 76, in select
    listeners.append(hub.add(hub.READ, k, on_read, current.throw, lambda: None))
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/kqueue.py", line 50, in add
    listener = super(Hub, self).add(evtype, fileno, cb, tb, mac)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 178, in add
    evtype, fileno, evtype, cb, bucket[fileno]))
RuntimeError: Second simultaneous read on fileno 5 detected.  Unless you really know what you're doing, make sure that only one greenthread can read any particular socket.  Consider using a pools.Pool. If you do know what you're doing and want to disable this error, call eventlet.debug.hub_prevent_multiple_readers(False) - MY THREAD=<function on_read at 0x10788e410>; THAT THREAD=FdListener('read', 5, <function on_read at 0x1071a3758>, <built-in method throw of greenlet.greenlet object at 0x10801b5f0>)

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@scipy.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True

Traceback (most recent call last):
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 458, in fire_timers
    timer()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/timer.py", line 58, in __call__
    cb(*args, **kw)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/semaphore.py", line 147, in _do_acquire
    waiter.switch()
error: cannot switch to a different thread
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/imris/anaconda/envs/test/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/core/history.py", line 852, in stop
    self.join()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/threading.py", line 940, in join
    self.__block.wait()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/threading.py", line 340, in wait
    waiter.acquire()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/semaphore.py", line 115, in acquire
    hubs.get_hub().switch()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 295, in switch
    return self.greenlet.switch()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 347, in run
    self.wait(sleep_time)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/kqueue.py", line 98, in wait
    time.sleep(seconds)
KeyboardInterrupt
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/Users/imris/anaconda/envs/test/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/IPython/core/history.py", line 852, in stop
    self.join()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/threading.py", line 940, in join
    self.__block.wait()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/threading.py", line 340, in wait
    waiter.acquire()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/semaphore.py", line 115, in acquire
    hubs.get_hub().switch()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 295, in switch
    return self.greenlet.switch()
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 347, in run
    self.wait(sleep_time)
  File "/Users/imris/anaconda/envs/test/lib/python2.7/site-packages/eventlet/hubs/kqueue.py", line 98, in wait
    time.sleep(seconds)
KeyboardInterrupt
```

I'm using OS X 10.11.6
Ipython=5.3.0

to recreate
```
conda create -n test ipython pip
source activate test
pip install eventlet  # I've also tried installing directly from github and got the same outcome
ipython
```

In Ipython:
```
import eventlet
eventlet.monkey_patch()
```

**Comment by temoto** [comment]
Unfortunately, it's a known issue without a good solution yet.

If possible, try `monkey_patch(os=False)`.

**Comment by isofer** [comment]
Thanks for the quick reply. 
I tried this solution but it did not help. :(

**Comment by temoto** [comment]
`monkey_patch(thread=False)` or both?

**Comment by isofer** [comment]
failed and failed.

**Comment by temoto** [comment]
Good news it reproduces on my machine.
Bad news everything I tried so far doesn't help.
Stay tuned.

**Comment by marco-lavagnino** [comment]
This happens on Linux too, I would remove the 'platform-osx' tag.

**Comment by marco-lavagnino** [comment]
I've been trying to reproduce this in a test, without luck :(

Manually testing, (installing ipython and eventlet in virtualenvs, monkey patching and seeing if it hangs) I can say that:

- This happens only for ipython 5.x (from 5.0.0b1 to 5.4.1)
- And it only happens when using python 2 (tested 2.7.12 and 2.7.13). Python 3.5.2 does not hang. 

Sadly, ipython 5 is the only supported version of the package compatible with python 2.7.

**Comment by temoto** [comment]
When this issue is resolved, put a notice at https://stackoverflow.com/questions/40842877

**Comment by Karthik777** [comment]
Silly question. But have you guys tried applying monkey patch before you import anything. and if my understanding is right. monkey patching doesn't work with debugging.

**Comment by deagon** [comment]
@marco-lavagnino 
In my tests, it still didn't work when changing the version of ipython and python interpreter.
I solved it by using the method `import_patched` instead of `monkey_patch` to decrease the impact of patches .

**Comment by temoto** [comment]
Sorry for leaving this issue open.

**Comment by codeSamuraii** [comment]
Hello,

Monkey patching the **select** module seems to be the reason IPython hangs.

Using `eventlet.monkey_patch(select=False)` fixes the issue for me.

**Comment by codeSamuraii** [comment]
**TLDR:** it might be necessary to exclude both **select and thread** from monkey patching to allow IPython to behave normally.

---
Update,

Not patching **select** allows IPython to continue running, however it hangs again when exiting.

https://github.com/eventlet/eventlet/assets/17270548/224b986d-f132-4c9a-ac76-0ce63e31aae3


Halting with `Ctrl-C` gives the following traceback:
```tb
Traceback (most recent call last):
  File "/Users/remi/.pyenv/versions/3.11.6/lib/python3.11/site-packages/IPython/core/history.py", line 923, in stop
    self.join()
  File "/Users/remi/.pyenv/versions/3.11.6/lib/python3.11/threading.py", line 1119, in join
    self._wait_for_tstate_lock()
  File "/Users/remi/.pyenv/versions/3.11.6/lib/python3.11/threading.py", line 1139, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt:
```
Not patching **select and thread** resolves all issues:

https://github.com/eventlet/eventlet/assets/17270548/c51b3864-21b4-4f79-954a-3517005babc0
